src/main/java/scala_maven/ScalaDocMojo.java

```diff
@@ -153,7 +154,14 @@ protected JavaMainCaller getScalaCommand() throws Exception {
     // This ensures we have a valid scala version...
     checkScalaVersion();
     Context sc = findScalaContext();
-    JavaMainCaller jcmd = getEmptyScalaCommand(sc.apidocMainClassName(scaladocClassName));
+    String apidocMainClassName = sc.apidocMainClassName(scaladocClassName);
+    JavaMainCaller jcmd;
+    if (sc.version().major < 3) {
+      jcmd = getEmptyScalaCommand(apidocMainClassName);
+    } else {
+      String targetClassesDir = project.getModel().getBuild().getOutputDirectory();
+      jcmd = new ScalaDoc3Caller(this, apidocMainClassName, targetClassesDir);
+    }
```

Comment:
The original purpose of Context (and findScalaContext) is to avoid / to remove every `if scalaVersion ...`  over the code. Can you try to keep this approach ?

Reply:
I tried to used getEmptyScalaCommand() but I couldn't get it working. In this case, the invocation of `dotty.tools.scaladoc.Main` raises a `NullPointerException`:
```
java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.getResourceAsStream(String)" because the return value of "java.lang.Class.getClassLoader()" is null
        at dotty.tools.scaladoc.renderers.Resources.renderResource(Resources.scala:213)
        at dotty.tools.scaladoc.renderers.Resources.renderResource$(Resources.scala:26)
        at dotty.tools.scaladoc.renderers.Renderer.renderResource(Renderer.scala:27)
        at dotty.tools.scaladoc.renderers.HtmlRenderer.renderResources$$anonfun$1(HtmlRenderer.scala:60)
        at scala.collection.immutable.List.flatMap(List.scala:293)
        at scala.collection.immutable.List.flatMap(List.scala:79)
        at dotty.tools.scaladoc.renderers.HtmlRenderer.renderResources(HtmlRenderer.scala:60)
        at dotty.tools.scaladoc.renderers.HtmlRenderer.render(HtmlRenderer.scala:31)
        at dotty.tools.scaladoc.Scaladoc$.run(Scaladoc.scala:249)
        at dotty.tools.scaladoc.Scaladoc$.run$$anonfun$1(Scaladoc.scala:84)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.Option.map(Option.scala:242)
        at dotty.tools.scaladoc.Scaladoc$.run(Scaladoc.scala:88)
        at dotty.tools.scaladoc.Main.run(Main.scala:18)
        at dotty.tools.scaladoc.Main$.main(Main.scala:24)
        at dotty.tools.scaladoc.Main.main(Main.scala)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at scala_maven_executions.MainHelper.runMain(MainHelper.java:157)
        at scala_maven_executions.MainWithArgsInFile.main(MainWithArgsInFile.java:30)
```
I tried to shift things from the bootclasspath to the normal classpath and some other things but without success. IMHO you should consider the removal of the Apache Commons Exec library in favor of `ProcessBuilder` and related classes from the JDK in the medium term. The Exec library hasn't seen any updates in years whereas `ProcessBuilder` and related classes have seen some improvements. I might be wrong about this but if someone upstream (Scala compiler etc.) decides that the usage of `java.lang.Class.getClassLoader()` might come in handy you might run into a similar issue as shown above.

The error above can be reproduced with the branch `classloader-issue` of the forked [repo](https://github.com/chr78rm/scala-maven-plugin/tree/classloader-issue). This branch isn't merged but `git diff master..classloader-issue` gives you the relevant changes.

Since we are raising external processes the outcome depends very much on how well the environment has been prepared. But, I don't think the classloader issue has something to do with the environment since the `ProcessBuilder` works well enough.

I have used the interface `JavaMainCaller` to hide the internal details of the process invocation.